### PR TITLE
SONARIAC-1200 S6865: Service account tokens should not be mounted in pods

### PR DESCRIPTION
**Pull request by navendu-barua-sonarsource**

### CODE REVIEW

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/checks/AbstractLimitsCheck.java [hunk 1]
```diff
@@ -39,14 +40,19 @@ void registerObjectCheck() {
...
   }
 
   @Nullable
-  static HasTextRange getFirstChildElement(BlockObject blockObject) {
+  static HasTextRange retrieveTextRangeToRaiseIssue(BlockObject blockObject) {
     if (blockObject.tree != null) {
-      return blockObject.tree.elements().get(0).key().metadata();
+      return blockObject.tree.elements().stream()
+        .filter(element -> element.key() instanceof ScalarTreeImpl)
+        .filter(element -> "spec".equals(((ScalarTreeImpl) element.key()).value()))
```
Comment (jonas-wielage-sonarsource) [inline iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/checks/AbstractLimitsCheck.java:52]: This implementation is highly specific to `spec` and does not fit with the theme of the `AbstractLimitCheck`. I would leave it as it is, and implement the required logic in the Check itself.

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/checks/AutomountServiceAccountTokenCheck.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.sonar.check.Rule;
+import org.sonar.iac.common.yaml.object.BlockObject;
+
+import static org.sonar.iac.common.yaml.TreePredicates.isSet;
+import static org.sonar.iac.common.yaml.TreePredicates.isTrue;
+import static org.sonar.iac.kubernetes.checks.AbstractLimitsCheck.retrieveTextRangeToRaiseIssue;
+
+@Rule(key = "S6865")
+public class AutomountServiceAccountTokenCheck extends AbstractKubernetesObjectCheck {
+  private static final String MESSAGE = "Set automountServiceAccountToken to false for the specification of kind %s.";
```
Comment (jonas-wielage-sonarsource) [inline iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/checks/AutomountServiceAccountTokenCheck.java:32]:
```suggestion
  private static final String MESSAGE = "Set automountServiceAccountToken to false for this specification of kind %s.";
```
Also please update rspec (I know, tedious, sorry), with this updated message.

**Review by jonas-wielage-sonarsource — COMMENTED** [review]
Nice PR, logic works quite good. I have some minor improvement ideas, please look at them.

File: iac-extensions/kubernetes/src/test/resources/checks/AutomountServiceAccountTokenCheck/test_pod_object.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  containers:
+    - name: nginx-compliant
+      image: nginx:1.25.3
+  automountServiceAccountToken: false # Compliant
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: tokenNotPresent
+spec: # Noncompliant
```
[the remark this answers is not in the document]
Comment (navendu-barua-sonarsource) [reply iac-extensions/kubernetes/src/test/resources/checks/AutomountServiceAccountTokenCheck/test_pod_object.yaml:26]: as `spec` starts at 0, it is not clear how I can add # and the highlight correctly

File: iac-extensions/kubernetes/src/test/resources/checks/AutomountServiceAccountTokenCheck/test_pod_object.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      image: nginx:1.25.3
+  automountServiceAccountToken: null # Noncompliant
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: tokenSetToTrueButContainersNotPresent
+spec:
+  automountServiceAccountToken: true # Compliant
```
[the remark this answers is not in the document]
Comment (navendu-barua-sonarsource) [reply iac-extensions/kubernetes/src/test/resources/checks/AutomountServiceAccountTokenCheck/test_pod_object.yaml:47]: yes

File: iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/checks/AutomountServiceAccountTokenCheck.java (hunk 2 above)
Comment (navendu-barua-sonarsource) [reply iac-extensions/kubernetes/src/main/java/org/sonar/iac/kubernetes/checks/AutomountServiceAccountTokenCheck.java:32]: created the PR for this

File: iac-extensions/kubernetes/src/test/resources/checks/AutomountServiceAccountTokenCheck/test_pod_object.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonas-wielage-sonarsource) [reply iac-extensions/kubernetes/src/test/resources/checks/AutomountServiceAccountTokenCheck/test_pod_object.yaml:26]: Haha you're right. As you have it in the other test file it's fine

**Review by jonas-wielage-sonarsource — APPROVED** [review]
LGTM, nice cleanup

**Pull request merged**